src/graph/optimizer/OptGroup.cpp

```diff
@@ -210,5 +225,28 @@ const PlanNode *OptGroupNode::getPlan() const {
   return node_;
 }
 
+void OptGroupNode::rebuildInputRelationship(const std::vector<OptGroup *> &boundary) {
+  // reach the boundary
+  if (std::all_of(dependencies_.begin(), dependencies_.end(), [&boundary](OptGroup *dep) {
+        return std::find(boundary.begin(), boundary.end(), dep) != boundary.end();
+      })) {
+    return;
+  }
+  if (std::find(boundary.begin(), boundary.end(), group_) != boundary.end()) {
+    return;
+  }
+  // Rebuild input
+  if (node_->inputVars().size() == dependencies_.size()) {
+    // Can't build input for some nodes, maybe need redesign it and corresponding rule.
+    for (std::size_t i = 0; i < dependencies_.size(); i++) {
+      const OptGroup *dep = dependencies_[i];
+      node_->setInputVar(dep->outputVar(), i);
+      for (auto *groupNode : dep->groupNodes()) {
```

Comment:
I think we need not to rebuild all `OptGroupNodes` in the depended `OptGroup` since there are many unchanged nodes which are not in the path of the OptRule. It will save much time to rebuild the nodes of transformed by the OptRule. WDYT?

Reply:
We could let users do it in rule, and only check it here. As you said before.